### DISCUSSION TITLE
Problem with Semantic Highlighting in Latest clangd

### DISCUSSION THREAD

**Discussion by daiyousei-qz** [clangd/clangd#1598] [Ideas]
In the current version of clangd, category of `Operator` is added to semantic tokens. Because of that, we cannot assign different operators with different color anymore. What's most bothering me is that, I'd like new/delete to be colored like a keyword while have other operators colored as operator.

Could we add some configurations to .clangd so we could force clangd not produce a certain HighlightingKind? Say, we could force clangd not to produce Operator semantic tokens so editor's fallback highlighting could be used.

**Comment by HighCommander4** [comment]
cc @ckandeler

Could you share what editor your use, and what is your mechanism for assigning different colors to different operators?

**Reply by daiyousei-qz** [reply]
vscode. Basically, by using the textmate rules.

I think the following quote should answer your question (Sorry, I edited the original post):

> What's most bothering me is that, I'd like new/delete to be colored like a keyword while have other operators still colored as operator.

**Reply by HighCommander4** [reply]
Ah, right, since the semantic token color always takes precedence over the client-side (textmate rules) color.

So one potential solution here could to be add a config option that disables sending operator tokens for all operators (perhaps only sending it for overloaded operators instead, which is an interesting piece of semantic information). This actually came up recently in the discussion of https://reviews.llvm.org/D143260, in the context of performance concerns about the volume of semantic tokens.

But I wonder if another solution to consider could be to just stop sending operator tokens for new/delete (just in general, not dependent on any configuration)? These are keywords (I mean that in the sense of "identifiable by their lexical token kinds, with no parsing required", not an opinion on how they should be colored), so the client should be able to easily identify and highlight them without the need for semantic tokens.

**Reply by daiyousei-qz** [reply]
I don't think such filtering could cause performance issues in comparison to adding more semantic tokens.

- For `HighlightingKind`, since it's an incremental enum, the filtering only requires an array lookup and one additional branch in `HighlightingsBuilder`.
- For `HighlightingModifier`, since it's a bit-enum. Only one additional masking is required.

So basically, the performance overhead is a couple more instructions per token, which should be trivial since we already spend a lot to collect those tokens anyway.

**Reply by daiyousei-qz** [reply]
On top of that, disabling a few token kinds could also reduce the number of total tokens to be sorted and sent back to clients. So filtering them could actually have performance benefits (for people who are actually using the config, though).

**Reply by HighCommander4** [reply]
To clarify, the perfomance concern is not with the filtering proposed here. The concern came up in the review of a different change, and it was about the amount of data we send over LSP in the semantic tokens message. Implementing the described filtering would actually help the performance, as you say.

**Reply by daiyousei-qz** [reply]
Thanks for the clarification! Let me work on a patch to achieve this. I think most of the change would be parsing the config file and pass things around.

Do you think I should open an issue for this?

**Reply by HighCommander4** [reply]
I mentioned two possible solution approaches above:

> add a config option that disables sending operator tokens for all operators (perhaps only sending it for overloaded operators instead

and

>  just stop sending operator tokens for new/delete (just in general, not dependent on any configuration)

Which one of these are you interested in writing a patch for?

I do think there is an advantage to a solution that does not require any configuration (i.e. the second one) over one that does (i.e. the first one). That said, I'd like to hear @ckandeler's feedback as well before proceeding, since he is the one who added the original operator semantic tokens.

**Reply by daiyousei-qz** [reply]
Oh, I’m looking forward to adding config to disable producing an arbitrary kind of highlight tokens, say operator, or bracket. Besides my need, I think it should be generally useful to allow user to customize what kinds of semantic tokens clangd produces. Also, it might also be useful as a gatekeeper if in the future we want to implement semantic tokens that’s disabled by default. My envisioned config looks like following:

SemanticTokens:
    DisableKinds: [ Operator, InactiveCode ]
    DisableModifiers: [ Readonly ]

**Reply by HighCommander4** [reply]
Ok, this approach seems nice and general, I like it.

**Comment by ckandeler** [comment]
I have of course no problem with such a config option.
I'm a bit confused, though: If the presence of the operator token kind causes the client to render things differently, doesn't this mean it has specific support for that and it is its own policy that causes that to happen?

**Reply by daiyousei-qz** [reply]
I’m not sure about other editor, but for vscode, if a standard semantic token type is not assigned with a theme, it would fallback to a syntactic theme.

The real problem here is the granularity. By introducing the operator token kind, we are now theming all operator tokens the same way. We cannot color ”new” and ”+” differently anymore, which sounds undesirable to me.

**Reply by HighCommander4** [reply]
To provide a bit more context: vscode's default themes provide a default color for every token kind in the LSP spec, including `operator`. In addition, vscode has client-side highlighting for things that can be recognized using a TextMate grammar rule. Semantic highlighting takes precedence over client-side highlighting.

The client-side highlightings are fairly granular, e.g. you can assign different colors to different operators (in the case of `new` vs. symbols like `+`, those may even be assigned different colors by default). If the server does not send semantic tokens for operators, it's these client-side highlight colors that are rendered. If the server does send semantic tokens for operators, those take precedence, and as a result e.g. `new` and `+` will now be colored the same.
